### PR TITLE
CUDA implementation 

### PR DESCRIPTION
**Pull request by tresreid**
Includes two cuda versions, one with unified memory and one without (v2). 
includes changes to the Makefile to run the cuda version
Minimal changes to openACC and openMP versions (get and print additional timing information).

### CODE REVIEW

**Comment by cerati** [comment]
I expected only new files, why there are changes to the Makefile and to other implementations?

**Comment by cerati** [comment]
I guess the printouts are fine, but you also remove some files. And the Makefile may mean that other versions won't compile?

**Comment by tresreid** [comment]
I only added some timing printouts for the acc and omp versions. The actual timing method (using the omp timer) will still have to be changed. 
I've added back the accidentally deleted files.
I've removed the additional flags I added in the makefile. I was told to add them to get the best performance for openMP. I guess you mean those options won't work on other architectures?  
Is that everything to fix? 
The other addition to the make file just includes an option to compile cuda. Also the target file includes the compiler and mode names so more than one binary can be made at once. Is that ok?

File: Makefile [hunk 1]
```diff
@@ -1,33 +1,35 @@
...
+MODE ?= cuda
 
 ######################################
 # Set the input source files (CSRCS) #
 ######################################
 ifeq ($(SRCTYPE),cpp)
 ifeq ($(MODE),acc)
-CSRCS = propagate-toz-test_OpenACC.cpp
+CSRCS = propagate-toz-test_OpenACC_v5.cpp
 else
```
Comment (cerati) [inline Makefile]: why do you change version here? and also at lines 24, 28, 32

Comment (tresreid) [reply Makefile]: ok. I've taken out the version numbers. However, I was under the impression that those were the lastest, fastest versions. Is that not the case? Those were the versions I had made comparisons to.

Comment (cerati) [reply Makefile]: If that is the case than the changes were fine... but I do not know for sure so I prefer to err on the conservative side. I am going to merge the PR, and we can update later. Thanks and sorry for being picky!

**Pull request merged**